src/tests/compiler/compiler.rs

```diff
@@ -1166,3 +1166,250 @@ fn test_inline_out_of_bounds_diagnostic() {
+      ()
+      )
+    )
+
+  (map
+    (let ((A (* add-number 2)))
+      (lambda ((& A) number) (+ A number))
+      )
+    L
+    )
+  )
+"}
+    .to_string();
+    let res = run_string(&prog, &"(5 (1 2 3 4))".to_string()).unwrap();
+    assert_eq!(res.to_string(), "(11 12 13 14)");
+}
+
+#[test]
+fn test_map_with_lambda_function_from_env_and_bindings() {
+    let prog = indoc! {"
+    (mod (add-number L)
+
+     (include *standard-cl-21*)
+
+     (defun map (F L)
+      (if L
+       (c (a F (list (f L))) (map F (r L)))
+       ()
+      )
+     )
+
+     (defun add-twice (X Y) (+ (* 2 X) Y))
+
+     (map
+      (lambda ((& add-number) number) (add-twice add-number number))
+      L
+     )
+    )"}
+    .to_string();
+    let res = run_string(&prog, &"(5 (1 2 3 4))".to_string()).unwrap();
+    assert_eq!(res.to_string(), "(11 12 13 14)");
+}
+
+#[test]
+fn test_map_with_lambda_function_from_env_no_bindings() {
+    let prog = indoc! {"
+    (mod (L)
+
+     (include *standard-cl-21*)
+
+     (defun map (F L)
+      (if L
+       (c (a F (list (f L))) (map F (r L)))
+       ()
+      )
+     )
+
+     (defun sum-list (L)
+       (if L
+         (+ (f L) (sum-list (r L)))
+         ()
+         )
+       )
+
+     (map
+      (lambda (lst) (sum-list lst))
+      L
+     )
+    )"}
+    .to_string();
+    let res = run_string(&prog, &"(((5 10 15) (2 4 8) (3 6 9)))".to_string()).unwrap();
+    assert_eq!(res.to_string(), "(30 14 18)");
+}
+
+#[test]
+fn test_lambda_using_let() {
+    let prog = indoc! {"
+    (mod (P L)
+
+     (include *standard-cl-21*)
+
+     (defun map (F L)
+      (if L
+       (c (a F (list (f L))) (map F (r L)))
+       ()
+      )
+     )
+
+     (map
+      (lambda ((& P) item) (let ((composed (c P item))) composed))
+      L
+     )
+    )"}
+    .to_string();
+    let res = run_string(&prog, &"(1 (10 20 30))".to_string()).unwrap();
+    assert_eq!(res.to_string(), "((1 . 10) (1 . 20) (1 . 30))");
+}
+
+#[test]
+fn test_lambda_using_macro() {
+    let prog = indoc! {"
+    (mod (P L)
+
+     (include *standard-cl-21*)
+
+     (defun map (F L)
+      (if L
+       (c (a F (list (f L))) (map F (r L)))
+       ()
+      )
+     )
+
+     (map
+      (lambda ((& P) item) (list P item))
+      L
+     )
+    )"}
+    .to_string();
+    let res = run_string(&prog, &"(1 (10 20 30))".to_string()).unwrap();
+    assert_eq!(res.to_string(), "((1 10) (1 20) (1 30))");
+}
+
+#[test]
+fn test_lambda_reduce() {
+    let prog = indoc! {"
+    (mod (LST)
+     (include *standard-cl-21*)
+     (defun reduce (fun lst init)
+      (if lst
+       (reduce fun (r lst) (a fun (list (f lst) init)))
+       init
+       )
+      )
+
+     (let
+      ((capture 100))
+      (reduce (lambda ((& capture) (X Y) ACC) (+ (* X Y) ACC capture)) LST 0)
+      )
+     )
+    "}
+    .to_string();
+    let res = run_string(&prog, &"(((2 3) (4 9)))".to_string()).unwrap();
+    assert_eq!(res.to_string(), "242");
+}
```

Comment:
I appreciate these tests. I think it would make sense to have negative tests as well. it's important compilation fails when it should

Reply:
It's been a bit hard because not much actually is meant to fail in chialisp and I often discover things people want that are questionable or should fail to me, but I will set those bounds as well as I can.